pkg/tfgen/convert_cli.go

```diff
@@ -0,0 +1,467 @@
+
+	// Convert all stubs populated by StartConvertingExamples.
+	fixedBytes := re.ReplaceAllFunc(bytes, func(match []byte) []byte {
+		groups := re.FindSubmatch(match)
+		i, err := strconv.Atoi(string(groups[1]))
+		contract.AssertNoErrorf(err, "strconv.Atoi")
+		ex := cc.convertExamplesList[i]
+		source := cc.generator.convertExamplesInner(ex.docs, ex.path,
+			ex.stripSubsectionsWithErrors, cc.generator.convertHCL)
+		// JSON-escaping to splice into JSON string literals.
+		bytes, err := json.Marshal(source)
+		contract.AssertNoErrorf(err, "json.Masrhal(sourceCode)")
+		return bytes[1 : len(bytes)-1]
+	})
+
+	var result pschema.PackageSpec
+	err = json.Unmarshal(fixedBytes, &result)
+	contract.AssertNoErrorf(err, "json.Unmarshal failed to recover PackageSpec")
+	return result
+}
+
+// During FinishConvertingExamples pass, generator calls back into this function to continue
+// PCL->lang translation from a pre-computed HCL->PCL translation table cc.pcls.
+func (cc *cliConverter) Convert(hclCode string, lang string) (string, hcl.Diagnostics, error) {
+	example, ok := cc.pcls[hclCode]
+	// Cannot assert here because panics are still possible for some reason.
+	// Example: gcp:gameservices/gameServerCluster:GameServerCluster
+	//
+	// Something skips adding failing conversion diagnostics to cc.pcls when pre-converting. The
+	// end-user experience is not affected much, the above example does not regress.
+	if !ok {
+		return "", hcl.Diagnostics{}, fmt.Errorf("unexpected HCL snippet in Convert")
+	}
+	if example.Diagnostics.HasErrors() {
+		return "", example.Diagnostics, nil
+	}
+	source, diags, err := cc.convertPCL(cc.currentPackageSpec, example.PCL, lang)
+	return source, cc.removeFileName(diags).Extend(example.Diagnostics), err
+}
+
+// Convert all observed HCL snippets from cc.hcls to PCL in one pass, populate cc.pcls.
+func (cc *cliConverter) bulkConvert() error {
+	examples := map[string]string{}
+	n := 0
+	for hcl := range cc.hcls {
+		fileName := fmt.Sprintf("e%d", n)
+		examples[fileName] = hcl
+		n++
+	}
+	result, err := cc.convertViaPulumiCLI(examples, []struct {
+		name string
+		info tfbridge.ProviderInfo
+	}{
+		{
+			name: cc.packageName,
+			info: cc.info,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	for fileName, hcl := range examples {
+		r := result[fileName]
+		cc.pcls[hcl] = translatedExample{
+			PCL:         r.PCL,
+			Diagnostics: cc.removeFileName(r.Diagnostics),
+		}
+	}
+	return nil
+}
+
+// Calls pulumi convert to bulk-convert examples.
+//
+// To facilitate high-throughput conversion, an `examples.json` protocol is employed to convert
+// examples in batches. See pulumi/pulumi-converter-terraform#29 for where the support is
+// introduced.
+//
+// Source examples are passed in as a map from ID to raw TF code.
+//
+// This may need to be coarse-grain parallelized to speed up larger providers at the cost of more
+// memory, for example run 4 instances of `pulumi convert` on 25% of examples each.
```

Comment:
I would prefer if this kind of speedup could happen in the engine, where we just pass the number of concurrent processes to run at once. It is safe to parallelize within process because there is no shared mutable state.[^1]

[^1]: If there is shared mutable state, then parallelizing by process splitting will be non-deterministic.

Reply:
>  It is safe to parallelize within process because there is no shared mutable state.

Famous last words. Prove there's no shared mutable state in large go systems (think static `var`).

Reply:
Hence the footnote: It will be either safe to parallelize within a single process or it isn't, in which case I'll be scared to parallelize across processes because each update shares mutable state.

Reply:
At a first approximation, yes, but it's not apples to apples. 

If you're parallelizing in-process, you can get issues related to say panics on concurrent map element access for shared state.

If you're parallelizing outside of process, you can get issues related to using shared OS resources such as file handles or folders.

It's just a guess at this point, I don't know if any of those are actual issues. We could try and see at some point.